src/internal/collection/postgres_collection.go

```diff
@@ -151,14 +183,44 @@ func (c *postgresCollection) get(ctx context.Context, key string, q sqlx.Queryer
 	return result, nil
 }
 
-func (c *postgresReadOnlyCollection) Get(key string, val proto.Message) error {
-	result, err := c.get(c.ctx, key, c.db)
+func (c *postgresReadOnlyCollection) Get(key interface{}, val proto.Message) error {
+	var result *model
+	var err error
+	err = c.withKey(key, func(rawKey string) error {
+		result, err = c.get(c.ctx, rawKey, c.db)
+		return err
+	})
 	if err != nil {
 		return err
 	}
 	return errors.EnsureStack(proto.Unmarshal(result.Proto, val))
 }
 
+func (c *postgresCollection) withKey(key interface{}, query func(string) error) error {
+	if str, ok := key.(string); ok {
+		return query(str)
+	}
+	rawKey, err := c.keyGen(key)
+	if err != nil {
+		return err
+	}
+	err = query(rawKey)
+	if err != nil {
+		var notFound ErrNotFound
+		var exists ErrExists
+		if c.notFound != nil && errors.As(err, &notFound) {
```

Comment:
I think you could get rid of this bit of complexity if the error lazily does the formatting (like instead of storing a custom message, store a reference to the formatter, then do this logic in the `ErrNotFound.Error` call).  And then always set the formatter to `c.notFound` or `c.exists` whenever those error types are constructed.

Reply:
hmm, I'm not sure where to pass the proto in, then

Reply:
I see, this is ~probably fine then.